plutus-contract-exe/src/Language/Plutus/Contract/State.hs

```diff
@@ -0,0 +1,324 @@
+
+To actually run a 'StatefulContract', we need to
+* use 'initialise' to get the initial state (of type 'Record') of the contract
+* for each new event, run 'insertAndUpdate' with the current state and the event
+  to get the new state. 
+
+TODO: There is a lot of boilerplate in the definitions of 'runClosed' and 
+'runOpen' - the cases we have to cover are all possible combinations of the
+constructors of 'StatefulContract' and 'Record'. 
+
+I did attempt to reduce this by using 'newtype Rec i a = Rec Either (OpenRecord i) (ClosedRecord i, a)' and variations but always ran into trouble. I would be 
+very grateful for suggestions on how to improve this situation.
+
+-}
+
+data StatefulContract a where
+    CMap :: (a' -> a) -> StatefulContract  a' -> StatefulContract  a
+    CAp :: StatefulContract  (a' -> a) -> StatefulContract  a' -> StatefulContract  a
+    CBind :: StatefulContract  a' -> (a' -> StatefulContract  a) -> StatefulContract  a
+
+    CContract :: C.ContractPrompt Maybe a -> StatefulContract  a
+    CJSONCheckpoint :: (Aeson.FromJSON a, Aeson.ToJSON a) => StatefulContract  a -> StatefulContract  a
+
+initialise
+    :: ( MonadWriter Hooks m )
+    => StatefulContract a
+    -> m (Either (OpenRecord Event) (ClosedRecord Event, a))
+initialise = \case
+    CMap f con -> fmap (fmap f) <$> initialise con
+    CAp conL conR -> do
+        l' <- initialise conL
+        r' <- initialise conR
+        case (l', r') of
+            (Left l, Left r)             -> pure $ Left (OpenBoth l r)
+            (Right (l, _), Left r)       -> pure $ Left (OpenRight l r)
+            (Left l, Right (r, _))       -> pure $ Left (OpenLeft l r)
+            (Right (l, f), Right (r, a)) -> pure $ Right (ClosedBin l r, f a)
+    CBind c f -> do
+        l <- initialise c
+        case l of
+            Left l' -> pure $ Left (OpenBind l')
+            Right (l', a) -> do
+                r <- initialise (f a)
+                case r of
+                    Left r'       -> pure $ Left $ OpenRight l' r'
+                    Right (r', b) -> pure $ Right (ClosedBin l' r', b)
+    CContract con -> do
+        r <- runConM mempty con
+        case r of
+            Nothing -> pure $ Left $ OpenLeaf mempty
+            Just a  -> pure $ Right (ClosedLeaf (FinalEvents mempty), a)
+    CJSONCheckpoint con -> do
+        r <- initialise con
+        case r of
+            Left _       -> pure r
+            Right (_, a) -> pure $ Right (jsonLeaf a, a)
+
+checkpoint :: (Aeson.FromJSON a, Aeson.ToJSON a) => StatefulContract  a -> StatefulContract  a
+checkpoint = CJSONCheckpoint
+
+prtty :: StatefulContract  a -> String
+prtty = \case
+    CMap _ c -> "cmap (" ++ prtty c ++ ")"
+    CAp l r -> "cap (" ++ prtty l ++ ") (" ++ prtty r ++ ")"
+    CBind l _ -> "cbind (" ++ prtty l ++  ") f"
+    CContract _ -> "ccontract"
+    CJSONCheckpoint j -> "json(" ++ prtty j ++ ")"
+
+instance Functor StatefulContract where
+    fmap f = \case
```

Comment:
Couldn't this just be `fmap f = CMap f`?

Reply:
Yeah for some reason I wanted to keep the number of `CMap`s small, but it doesn't make a difference.